docs: Add e2e testing with dappwright (#536)
oasisprotocol/sapphire-paratime: docs/develop/dappwright.md

  },

  wallet: async ({ context }, use) => {
    const metamask = await dappwright.getWallet('metamask', context)

    await use(metamask)
  },
})
...
```

The above includes the Sapphire Localnet as a network with the correct RPC
for testing, and sets up the default MetaMask wallet to use the same [seed] as
you would in a Hardhat test.

[seed]: https://hardhat.org/hardhat-network/docs/reference#accounts

## Writing a Test

Writing a test with dappwright is the same as you would write a Playwright
test. We would want to visit our application first:

```typescript title="tests/e2e.spec.ts"
test.beforeEach(async ({ page }) => {
  await page.goto('http://localhost:5173')
})
```

Next, we can load the application and confirm using the Sapphire network in
Metamask. Note that **we will need to use `wallet.page` to access the MetaMask
extension and `wallet.page.reload()` while waiting for the MetaMask dom to
reload.**

```typescript title="tests/e2e.spec.ts"
test('set and view message', async ({ wallet, page }) => {
  // Load page
  await page.getByTestId('rk-connect-button').click()
  await page.getByTestId('rk-wallet-option-injected-sapphire').click()
  await wallet.page.reload()

>   await wallet.page.waitForLoadState('domcontentloaded')

Author: Test out `wallet.approve`

---
description: End-to-End Testing of Oasis Sapphire dApps
---

# End-to-End Testing

Many modern web applications use [Playwright] tests during the development and

> release process to crease shipping speed and improve quality. While dApps have

Reviewer: ```suggestion
release process to increase shipping speed and improve quality. While dApps have
```

a younger ecosystem, tools exist to do the same. We recommend using

> [dappwright] for dApps on the Sapphire Network. In this tutorial, we will

Reviewer: ```suggestion
[dAppwright] for dApps on the Sapphire Network. In this tutorial, we will
```

Their github has the A capitalized, I think we should keep it the same in our docs.

> breakdown the e2e testing involved in the [demo-starter] project.

Reviewer: ```suggestion
break down the e2e testing involved in the [demo-starter] project.
```

[Playwright]: https://playwright.dev/docs/intro

> [dappwright]: https://github.com/TenKeyLabs/dappwright

Reviewer: ```suggestion
[dAppwright]: https://github.com/TenKeyLabs/dappwright
```

[demo-starter]: https://github.com/oasisprotocol/demo-starter

> ## dappwright

Reviewer: ```suggestion
## dAppwright
```

> The [dappwright package] builds on Playwright and includes tooling to support

Reviewer: ```suggestion
The [dAppwright package] builds on Playwright and includes tooling to support
```

testing with a MetaMask or Coinbase wallet as an extension on a Chromium
browser.

> [dappwright package]: https://www.npmjs.com/package/@tenkeylabs/dappwright

Reviewer: ```suggestion
[dAppwright package]: https://www.npmjs.com/package/@tenkeylabs/dappwright
```

## Installation

> We need to install both `dappwright` and `Playwright`. Navigate to your

Reviewer: ```suggestion
We need to install both `dAppwright` and `Playwright`. Navigate to your
```

frontend application directory:

> 1. Install dappwright:

Reviewer: ```suggestion
1. Install dAppwright:
```

   ```shell npm2yarn
   npm install -D @tenkeylabs/dappwright
   ```

2. Install Playwright (we recommend the TypeScript option):

   ```shell npm2yarn
   npm init playwright@latest
   ```

3. A successful installation should allow the running of the example tests:

   ```shell
   npx playwright test
   ```

## Setup

We suggest starting a local dev server with each test run to iterate over the
same state.

```typescript title="playwright.config.ts"
import { defineConfig } from '@playwright/test';

export default defineConfig({
  // highlight-start
  /* Run your local dev server before starting the tests */
  webServer: {
    command: 'pnpm dev',
    url: process.env.FRONTEND_URL || 'http://localhost:8080/',
    reuseExistingServer: !process.env.CI,
    stdout: 'pipe',
    stderr: 'pipe',
  },
  // highlight-end
});
```

## Adding Test Context

> We begin with a test file extending the testing context to include dappwright:

Reviewer: ```suggestion
We begin with a test file extending the testing context to include dAppwright:
```

export const test = baseTest.extend<{
  context: BrowserContext
  wallet: Dappwright
}>({
  context: async ({}, use) => {
    // Launch context with extension
    const [wallet, _, context] = await dappwright.bootstrap('', {
      wallet: 'metamask',
      version: MetaMaskWallet.recommendedVersion,
      seed: 'test test test test test test test test test test test junk', // Hardhat's default https://hardhat.org/hardhat-network/docs/reference#accounts
      headless: false,
    })

    // Add Sapphire Localnet as a custom network
    await wallet.addNetwork({
      networkName: 'Sapphire Localnet',
      rpc: 'http://localhost:8545',
      chainId: 23293,
      symbol: 'ROSE',
    })

    await use(context)
  },

  wallet: async ({ context }, use) => {
    const metamask = await dappwright.getWallet('metamask', context)

    await use(metamask)
  },
})
...
```

The above includes the Sapphire [Localnet] as a network with the correct RPC
for testing, and sets up the default MetaMask wallet to use the same [seed] as
you would in a Hardhat test.

[seed]: https://hardhat.org/hardhat-network/docs/reference#accounts

## Writing a Test

> Writing a test with dappwright is the same as you would write a Playwright

Reviewer: ```suggestion
Writing a test with dAppwright is the same as you would write a Playwright
```

  await page.getByTestId('rk-wallet-option-injected-sapphire').click()
  await wallet.approve()
})
```

Otherwise, we write selectors and assertions in the same way.

```typescript title="tests/e2e.spec.ts"
  // Set a message
  await page.locator(':text-matches("0x.{40}")').fill('hola amigos')
  const submitBtn = page.getByRole('button', { name: 'Set Message' })
  await submitBtn.click()
  await wallet.confirmTransaction()

  // Reveal the message
  await expect(submitBtn).toBeEnabled()
  await page.locator('[data-label="Tap to reveal"]').click()
  await wallet.confirmTransaction()

  // Assert message has been set
  await expect(page.locator('[data-label="Tap to reveal"]').locator('input')).toHaveValue('hola amigos')
```

You can make assertions in the same way on the wallet page, but
[wallet actions] will significantly simplify the amount of boilerplate testing
code.

```typescript
  expect(wallet.page.getByText("My Account Name"));
```

[wallet actions]: https://github.com/TenKeyLabs/dappwright/blob/d791017c51edc4e61e786504c154f9ad3db43ab6/src/wallets/wallet.ts#L27-L45

## Debugging

Playwright's UI mode is very beneficial to debugging your tests as you develop.

> The pick locator button will you refine element selectors while giving visual

Reviewer: ```suggestion
The pick locator button will help you refine element selectors while giving visual
```

feedback.

```sh
npx playwright test --ui
```

Otherwise, you could consider the debug mode which includes breakpoints and
more complete interactivity of browser network requests.

```sh
npx playwright test --debug
```

## CI

> Running your dappwright tests on CI environments such as GitHub is possible

Reviewer: ```suggestion
Running your dAppwright tests on CI environments such as GitHub is possible
```

with the correct configurations. You will need to install `playwright` itself
as a dependency before you can install Playwright's dependency packages, and
run a [headed] execution in Linux agents with `Xvfb`. We recommend uploading
test results on failure to more quickly move through CI cycles.

You will need a Sapphire [Localnet] service to provide an RPC endpoint during
testing.

```yaml
  playwright-test:
    runs-on: ubuntu-latest
    // highlight-start
    services:
      sapphire-localnet-ci:
        image: ghcr.io/oasisprotocol/sapphire-localnet:latest
        ports:
          - 8545:8545
          - 8546:8546
        env:
          OASIS_DOCKER_START_EXPLORER: no
        options: >-
          --rm
          --health-cmd="test -f /CONTAINER_READY"
          --health-start-period=90s
    // highlight-end
```

We recommend saving deployed smart contract addresses as environment variables

> and [pass] `$GITHUB_OUTPUT` to a subsequent testing step.

Reviewer: ```suggestion
and [passing] `$GITHUB_OUTPUT` to a subsequent testing step.
```

```yaml
      - name: Deploy backend
        working-directory: backend
        id: deploy
        run: |
          echo "message_box_address=$(pnpm hardhat deploy localhost --network sapphire-localnet | grep -o '0x.*')" >> $GITHUB_OUTPUT
```

Finally, run the test and upload results on failure:

```yaml
      - name: Build
        working-directory: frontend
        run: pnpm build

      - name: Install Playwright dependencies
        run: pnpm test:setup
        working-directory: frontend

      - name: Run playwright tests (with xvfb-run to support headed extension test)
        working-directory: frontend
        run: xvfb-run pnpm test
        env:
          VITE_MESSAGE_BOX_ADDR: ${{ steps.deploy.outputs.message_box_address }}

      - name: Upload playwright test-results
        if: ${{ failure() }}
        uses: actions/upload-artifact@v4
        with:
          name: playwright-test-results
          path: frontend/test-results
          retention-days: 5
```

[headed]: https://playwright.dev/docs/ci#running-headed
[Localnet]: https://github.com/oasisprotocol/docs/blob/main/docs/build/tools/localnet.mdx

> [pass]: https://docs.github.com/en/actions/writing-workflows/choosing-what-your-workflow-does/passing-information-between-jobs

Reviewer: ```suggestion
[passing]: https://docs.github.com/en/actions/writing-workflows/choosing-what-your-workflow-does/passing-information-between-jobs
```

:::info Example: demo-starter

> If you are interested in seeing how dappwright is integrated into an example

Reviewer: ```suggestion
If you are interested in seeing how dAppwright is integrated into an example
```

application, check out the [demo starter].

> [demo starter]: https://github.com/oasisprotocol/demo-starter/tree/master/frontend

Author: Oh yes.

:::

:::info Example: wagmi

> If you are interested in seeing how dappwright is integrated into a Sapphire

Reviewer: ```suggestion
If you are interested in seeing how dAppwright is integrated into a Sapphire
```
